src/disposable/AsyncDisposable.ts

```diff
@@ -0,0 +1,42 @@
+import { IAsyncDisposable } from './types';
+
+export function createAsyncDisposable(func: () => void | Promise<void>): IAsyncDisposable {
+  return {
+    dispose: func,
+  };
+}
+
+export class AsyncDisposableCollection implements IAsyncDisposable {
+  protected readonly disposables: IAsyncDisposable[] = [];
+
+  get disposed(): boolean {
+    return this.disposables.length === 0;
+  }
+
+  public async dispose(): Promise<void> {
+    if (this.disposed) return;
+
+    while (!this.disposed) {
+      await this.disposables[this.disposables.length - 1].dispose();
+    }
+  }
+
+  public push(disposable: IAsyncDisposable): IAsyncDisposable {
+    this.disposables.push(disposable);
+
+    const originalDispose = disposable.dispose.bind(disposable);
+    const toRemove = createAsyncDisposable(() => {
+      const index = this.disposables.indexOf(disposable);
+      if (index !== -1) {
+        this.disposables.splice(index, 1);
+      }
+    });
+
+    disposable.dispose = async () => {
+      await originalDispose();
+      await toRemove.dispose();
+    };
+
+    return toRemove;
```

Comment:
Hang on. @P0lip do we actually use this `toRemove` disposable anywhere? I'm examing the original implementation closely and `.push` returns an `IDisposable` that removes it from the DisposableCollection but it _doesn't dispose the original disposable_. It's like an undo, "oops I don't want this Disposable to be managed by the DisposableCollection anymore" operation. But that _only_ is possibly useful if you still have the _original_ disposable around so you can dispose it manually.

Reply:
I think I'm gonna re-write the Disposable classes. I don't trust <!--Marc's--> the current implementations, they are too subtly complicated.